File: haskell/day_3/day3.hs
import System.IO
import Control.Monad
import Data.List
import Prelude

alphabet = ['a'..'z'] ++ ['A'..'Z']

compare1 x y = sum [i*j| (i,j) <- zip common [0..51]]
            where common = [i*j | (i,j) <- zip letters_x letters_y]
                         where letters_x = [if elem i x then 1 else 0 | i <- alphabet]
                               letters_y = [if elem i y then 1 else 0 | i <- alphabet]

split_two x = [[x !! i | i <- [0..half-1] ], [x !! i | i <- [half..total]]]
            where half = div (length x) 2
                  total = half * 2 - 1

solution x = compare1 r p
           where r = (split_two x) !! 0
                 p = (split_two x) !! 1
main = do
    handle <- openFile "input" ReadMode
    contents <- hGetContents handle
    print $  sum [solution x + 1 | x<-words contents]

File: python/day3.py
f = open("input",'r')
words = f.readlines()
ans = list()
alphabet = [chr(i) for i in range(97, 97+26)] + [chr(i) for i in range(65, 65+26)]

def compare(x,y):
    letters_y = [0 for _ in alphabet]
    letters_x = [0 for _ in alphabet]
    for i in y:
        for index, p in enumerate(alphabet):
            if p == i:
                letters_y[index] = 1
                break
    for i in x:
        for index, p in enumerate(alphabet):
            if p == i:
                letters_x[index] = 1
                break
    common = [i*j for i,j in zip(letters_y, letters_x)]
    return sum([i*j for i,j in enumerate(common)])
for w in words:
    half = int(len(w) / 2)
    comp1 = w[:half]
    comp2 = w[half:]
    index = compare(comp1,comp2)
    ans.append(index+1)
print(sum(ans))
